rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/action/TbSaveToCustomCassandraTableNodeTest.java

```diff
@@ -0,0 +1,323 @@
+                .isInstanceOf(RuntimeException.class)
+                .hasMessage("Message data doesn't contain key: 'temp'!");
+    }
+
+    @Test
+    void givenUnsupportedData_whenOnMsg_thenThrowsException() throws TbNodeException {
+        config.setTableName("temperature_sensor");
+        config.setFieldsMapping(Map.of("temp", "temperature"));
+
+        mockOnInit();
+        mockBoundStatement();
+
+        node.init(ctxMock, new TbNodeConfiguration(JacksonUtil.valueToTree(config)));
+
+        String data = """
+                {
+                "temp": [value]
+                }
+                """;
+        TbMsg msg = TbMsg.newMsg(TbMsgType.POST_TELEMETRY_REQUEST, DEVICE_ID, TbMsgMetaData.EMPTY, data);
+        assertThatThrownBy(() -> node.onMsg(ctxMock, msg))
+                .isInstanceOf(RuntimeException.class)
+                .hasMessage("Message data key: 'temp' with value: '[\"value\"]' is not a JSON Object or JSON Primitive!");
+    }
+
+    @Test
+    void givenValidMsgStructureAndKeyIsEntityId_whenOnMsg_thenOk() throws Exception {
+        config.setTableName("readings");
+        config.setFieldsMapping(Map.of("$entityId", "entityId"));
+
+        mockOnInit();
+        when(boundStatementBuilderMock.setUuid(anyInt(), any(UUID.class))).thenReturn(boundStatementBuilderMock);
+        when(boundStatementMock.getConsistencyLevel()).thenReturn(DefaultConsistencyLevel.ONE);
+        mockBoundStatementBuilder();
+        mockSubmittingCassandraTask();
+
+        node.init(ctxMock, new TbNodeConfiguration(JacksonUtil.valueToTree(config)));
+
+        TbMsg msg = TbMsg.newMsg(TbMsgType.POST_TELEMETRY_REQUEST, DEVICE_ID, TbMsgMetaData.EMPTY, TbMsg.EMPTY_JSON_OBJECT);
+        node.onMsg(ctxMock, msg);
+
+        verify(sessionMock).prepare("INSERT INTO cs_tb_readings(entityId) VALUES(?)");
+        verify(boundStatementBuilderMock).setUuid(anyInt(), eq(DEVICE_ID.getId()));
+
+        ArgumentCaptor<CassandraStatementTask> taskCaptor = ArgumentCaptor.forClass(CassandraStatementTask.class);
+        verify(ctxMock).submitCassandraWriteTask(taskCaptor.capture());
+        CassandraStatementTask task = taskCaptor.getValue();
+        assertThat(task.getTenantId()).isEqualTo(TENANT_ID);
+        assertThat(task.getSession()).isEqualTo(sessionMock);
+        assertThat(task.getStatement()).isEqualTo(boundStatementMock);
+        await().atMost(1, TimeUnit.SECONDS).untilAsserted(
+                () -> verify(ctxMock).tellSuccess(msg)
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource
+    void givenValidMsgStructure_whenOnMsg_thenSaveToCustomCassandraTable(
+            String valueFromData,
+            Consumer<BoundStatementBuilder> mockBuilderConsumer,
+            Consumer<BoundStatementBuilder> verifyBuilderConsumer) throws TbNodeException {
+        config.setTableName("sensor_readings");
+        config.setFieldsMapping(Map.of("msgField1", "tableColumn1", "msgField2", "tableColumn2"));
```

Comment:
why we need msgField2 in this test?  Is it used in the test somehow? Maybe it's better to add as many fields as you supported data types and replace the parameterized test then?